(PE-14979) add i18n support

jonathannewman opened the pull request:
This adds the clj-i18n library and translates all the strings in it
using the library.  In addition, this adds a test to validate that
the translations work.

aperiodic commented on locales/messages.pot:
```diff
@@ -0,0 +1,53 @@
+
+#: src/puppetlabs/jdbc_util/pool.clj:108
+msgid "Initialization resulted in an error: {0}"
+msgstr ""
```
If this is generated by the Makefile then why are we checking it in? Seems like the same thing as checking in the built binary, or CSS generated from SASS, both things we'd never do.

jonathannewman replied:
The translators need this file to perform the translation.  It is consumed by the translation tools and  .pot files will be added to the repo with the given translations.

aperiodic replied:
But if we have a Makefile that can just generate this, then why should we check it in? I'm okay with checking in the translated files, because we'll obviously need to, but I don't see why we should check in a generated file.

steveax replied:
@aperiodic, yeah, known thing, Libby noted this in [doc for the collab design meeting](https://docs.google.com/document/d/1eoOtB0EmsbjiPBz4lbhmlCG006CR9RBYgq7ZmTqZR98/edit#). Haven't come up with a solution for this yet.

aperiodic replied:
@steveax yeah, and I remember at the meeting that the general feeling was that we shouldn't check them in. Now @jonathannewman tells me "it's been decided" that the standard workflow is to check them in, and I don't understand how that decision was reached.

aperiodic commented on project.clj:
```diff
@@ -13,14 +13,17 @@
                                   [log4j/log4j "1.2.17"]]}}
 
-  :plugins [[lein-release "1.0.5"]]
+  :plugins [[org.clojure/clojure "1.8.0"]
```
wat

jonathannewman replied:
This is needed to fix a version conflict between lein-release and i18n.  Note that this same version is applied to the whole project so it seemed pretty safe.

aperiodic replied:
Ok, can you add a comment explaining why it's there?

jonathannewman replied:
Good idea, will do.

aperiodic commented on project.clj:
```diff
@@ -13,14 +13,17 @@
-  :plugins [[lein-release "1.0.5"]]
+  :plugins [[org.clojure/clojure "1.8.0"]
+            [lein-release "1.0.5"]
+            [puppetlabs/i18n "0.4.0"]]
```
Why is this both a dependency and a plugin?

jonathannewman replied:
See the i18n repo... it adds to the lein command set.

aperiodic commented on src/puppetlabs/jdbc_util/core.clj:
```diff
@@ -139,7 +140,9 @@
             replaced (str/replace match #"(.*)\?$" (str "$1" replacement))]
         (str replaced tail))
-      (throw (IllegalArgumentException. (str "there are not " n " '?'s in the given string"))))))
+      (throw (IllegalArgumentException. (trun "There is not one '?' in the given string"
```
I think the intent here would be a little clearer if the singular version were "There are no '?'s in the given string"

aperiodic replied:
I'm also not sure whether this counts as "user-facing"

jonathannewman replied:
Not clear how/where this exception will be handled, but seemed more likely to be something a user might see.  I agree that it isn't terribly important in the grand scheme of things.  I can change it to be more simple.

aperiodic replied:
This is prooobably a code bug but could not be. So I'm fine keeping it in but the singular case only happens when the string is supposed to contain a "?" and doesn't have any at all, so saying it doesn't have any at all gets the point across better.

aperiodic commented on test/puppetlabs/jdbc_util/Messages_eo.properties:
```diff
@@ -0,0 +1 @@
+Initialization\ resulted\ in\ an\ error\:\ {0}=This_is_a_translated_string: {0}
```
Missing a newline before EOF here.

jonathannewman replied:
Property files are a little magic.  I'll add it and make sure it doesn't blow up.  If it does, I'll remove it. :)

jonathannewman commented:
Added newline to properties file, changed singular translation for exception and modified it to be a system translation rather than a user translation.

jonathannewman commented:
Added a comment for the project.clj plugins addition.

aperiodic commented on src/puppetlabs/jdbc_util/core.clj:
```diff
@@ -139,7 +140,9 @@
             replaced (str/replace match #"(.*)\?$" (str "$1" replacement))]
         (str replaced tail))
-      (throw (IllegalArgumentException. (str "there are not " n " '?'s in the given string"))))))
+      (throw (IllegalArgumentException. (trsn "There is are no '?' in the given string"
```
Typo here: "is are no '?'".

aperiodic commented:
@jonathannewman there's a typo in the new singular replace-nth-? exception message.

jonathannewman commented:
@aperiodic Good catch.  Fixed it.

aperiodic commented on src/puppetlabs/jdbc_util/core.clj:
```diff
@@ -139,7 +140,9 @@
             replaced (str/replace match #"(.*)\?$" (str "$1" replacement))]
         (str replaced tail))
-      (throw (IllegalArgumentException. (str "there are not " n " '?'s in the given string"))))))
+      (throw (IllegalArgumentException. (trsn "There is are no '?'s in the given string"
```
@jonathannewman typo is still here: `There is are no '?'s`

jonathannewman replied:
Ok, really fixed now. :)

jonathannewman commented:
Address grammar issue.

aperiodic commented:
I'm okay with everything here, except checking in the `messages.pot` file. Being a generated file, I don't see why we would want to check it in, and checking it in comes with real drawbacks: because it changes every time a line is added or removed before a translated string in any file, almost any changes to the source will cause changes to the file. It adds a cost to nearly every future PR that all developers will have to pay; though small, I'd prefer to just avoid it. It will also make any future rebases & merge ups very tedious.

We've talked a little about this in person and not come to agreement; do you want to take it to a development list discussion?

pcarlisle commented:
@libbymo is working on a different approach to the `messages.pot` file that we were going to trial in a different project. I'd like to let that run a little before opening it up to full mailing list discussion.

aperiodic commented:
I guess we'll address the `messages.pot` issue in another forum; it's a bigger issue than just this project.
